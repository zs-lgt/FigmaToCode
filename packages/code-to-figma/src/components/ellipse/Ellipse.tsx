import * as React from 'react';
import { DefaultShapeProps, CornerProps, StyleOf, InstanceItemProps, SelectionEventProps } from '../../types';
import {
    LayoutStyleProperties,
    transformLayoutStyleProperties
} from '../../styleTransformers/transformLayoutStyleProperties';
import { useYogaLayout } from '../../hooks/useYogaLayout';
import { transformBlendProperties, BlendStyleProperties } from '../../styleTransformers/transformBlendProperties';
import {
    GeometryStyleProperties,
    transformGeometryStyleProperties
} from '../../styleTransformers/transformGeometryStyleProperties';
import { YogaStyleProperties } from '../../yoga/YogaStyleProperties';
import { StyleSheet } from '../..';
import { useSelectionChange } from '../../hooks/useSelectionChange';
import { transformAutoLayoutToYoga } from '../../styleTransformers/transformAutoLayoutToYoga';
import { OnLayoutHandlerProps, useOnLayoutHandler } from '../../hooks/useOnLayoutHandler';
import { useImageHash } from '../../hooks/useImageHash';
import { useNodeIdCallback } from '../../hooks/useNodeIdCallback';

export interface EllipseProps
    extends DefaultShapeProps,
        CornerProps,
        InstanceItemProps,
        SelectionEventProps,
        OnLayoutHandlerProps {
    style?: StyleOf<YogaStyleProperties & LayoutStyleProperties & GeometryStyleProperties & BlendStyleProperties>;
    children?: undefined;
    arcData?: ArcData;
}

const Ellipse: React.FC<EllipseProps> = props => {
    const nodeRef = React.useRef();
    useSelectionChange(nodeRef, props);
    useNodeIdCallback(nodeRef, props.onNodeId);
    const style = { ...StyleSheet.flatten(props.style), ...transformAutoLayoutToYoga(props) };

    const imageHash = useImageHash(style.backgroundImage);

    const ellipseProps = {
        ...transformLayoutStyleProperties(style),
        ...transformBlendProperties(style),
        ...transformGeometryStyleProperties('fills', style, imageHash),
        ...props,
        style
    };
    const yogaChildProps = useYogaLayout({ nodeRef, ...ellipseProps });
    useOnLayoutHandler(yogaChildProps, props);

    return <ellipse {...ellipseProps} {...yogaChildProps} innerRef={nodeRef} />;
};

export { Ellipse };
