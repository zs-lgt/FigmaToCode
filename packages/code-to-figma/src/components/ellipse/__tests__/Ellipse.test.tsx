import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { Ellipse } from '../Ellipse';
import { createFigma } from 'figma-api-stub';

describe('<Ellipse />', () => {
    beforeEach(() => {
        // @ts-ignore
        global.figma = createFigma({
            simulateErrors: true
        });
    });

    it('Ellipse without props', () => {
        const tree = renderer.create(<Ellipse />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('Ellipse with layout style properties', () => {
        const tree = renderer
            .create(
                <Ellipse
                    style={{
                        width: 100,
                        height: 200
                    }}
                />
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('Ellipse with arc data', () => {
        const tree = renderer
            .create(
                <Ellipse
                    arcData={{
                        startingAngle: 0,
                        endingAngle: Math.PI,
                        innerRadius: 0.5
                    }}
                />
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('Ellipse with color background', () => {
        const tree = renderer
            .create(
                <Ellipse
                    style={{
                        backgroundColor: '#1bff00'
                    }}
                />
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('Ellipse with image background', () => {
        const tree = renderer
            .create(
                <Ellipse
                    style={{
                        backgroundColor: '#ff8b29',
                        backgroundSize: 'cover',
                        backgroundImage: './image.png'
                    }}
                />
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
});
