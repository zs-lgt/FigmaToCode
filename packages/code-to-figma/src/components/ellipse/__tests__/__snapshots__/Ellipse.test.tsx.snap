// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Ellipse /> Ellipse with arc data 1`] = `
<ellipse
  arcData={
    Object {
      "endingAngle": 3.141592653589793,
      "innerRadius": 0.5,
      "startingAngle": 0,
    }
  }
  innerRef={
    Object {
      "current": undefined,
    }
  }
  style={Object {}}
/>
`;

exports[`<Ellipse /> Ellipse with color background 1`] = `
<ellipse
  fills={
    Array [
      Object {
        "color": Object {
          "b": 0,
          "g": 1,
          "r": 0.10588235294117647,
        },
        "opacity": 1,
        "type": "SOLID",
      },
    ]
  }
  innerRef={
    Object {
      "current": undefined,
    }
  }
  style={
    Object {
      "backgroundColor": "#1bff00",
    }
  }
/>
`;

exports[`<Ellipse /> Ellipse with image background 1`] = `
<ellipse
  fills={
    Array [
      Object {
        "color": Object {
          "b": 0.1607843137254902,
          "g": 0.5450980392156862,
          "r": 1,
        },
        "opacity": 1,
        "type": "SOLID",
      },
    ]
  }
  innerRef={
    Object {
      "current": undefined,
    }
  }
  style={
    Object {
      "backgroundColor": "#ff8b29",
      "backgroundImage": "./image.png",
      "backgroundSize": "cover",
    }
  }
/>
`;

exports[`<Ellipse /> Ellipse with layout style properties 1`] = `
<ellipse
  height={200}
  innerRef={
    Object {
      "current": undefined,
    }
  }
  style={
    Object {
      "height": 200,
      "width": 100,
    }
  }
  width={100}
/>
`;

exports[`<Ellipse /> Ellipse without props 1`] = `
<ellipse
  innerRef={
    Object {
      "current": undefined,
    }
  }
  style={Object {}}
/>
`;
